src/graphql_request_template.rs

```diff
@@ -0,0 +1,217 @@
+#![allow(clippy::too_many_arguments)]
+
+use derive_setters::Setters;
+use hyper::HeaderMap;
+use reqwest::header::{HeaderName, HeaderValue};
+
+use crate::config::{GraphQLOperationType, KeyValues};
+use crate::has_headers::HasHeaders;
+use crate::http::Method::POST;
+use crate::lambda::GraphQLOperationContext;
+use crate::mustache::Mustache;
+use crate::path::PathGraphql;
+
+/// RequestTemplate for GraphQL requests (See RequestTemplate documentation)
+#[derive(Setters, Debug, Clone)]
+pub struct GraphqlRequestTemplate {
+  pub url: String,
+  pub operation_type: GraphQLOperationType,
+  pub operation_name: String,
+  pub operation_arguments: Option<Vec<(String, Mustache)>>,
+  pub headers: Vec<(HeaderName, Mustache)>,
+}
+
+impl GraphqlRequestTemplate {
+  fn create_headers<C: PathGraphql>(&self, ctx: &C) -> HeaderMap {
+    let mut header_map = HeaderMap::new();
+
+    for (k, v) in &self.headers {
+      if let Ok(header_value) = HeaderValue::from_str(&v.render_graphql(ctx)) {
+        header_map.insert(k, header_value);
+      }
+    }
+
+    header_map
+  }
+
+  fn set_headers<C: PathGraphql + HasHeaders>(&self, mut req: reqwest::Request, ctx: &C) -> reqwest::Request {
+    let headers = req.headers_mut();
+    let config_headers = self.create_headers(ctx);
+
+    if !config_headers.is_empty() {
+      headers.extend(config_headers);
+    }
+    headers.insert(
+      reqwest::header::CONTENT_TYPE,
+      HeaderValue::from_static("application/json"),
+    );
+    headers.extend(ctx.headers().to_owned());
+    req
+  }
+
+  pub fn to_request<C: PathGraphql + HasHeaders + GraphQLOperationContext>(
+    &self,
+    ctx: &C,
+  ) -> anyhow::Result<reqwest::Request> {
+    let mut req = reqwest::Request::new(POST.to_hyper(), url::Url::parse(self.url.as_str())?);
+    req = self.set_headers(req, ctx);
+    req = self.set_body(req, ctx);
+    Ok(req)
+  }
+
+  fn set_body<C: PathGraphql + HasHeaders + GraphQLOperationContext>(
+    &self,
+    mut req: reqwest::Request,
+    ctx: &C,
+  ) -> reqwest::Request {
+    let operation_type = &self.operation_type;
+    let selection_set = ctx.selection_set().unwrap_or_default();
+    let operation = self
+      .operation_arguments
+      .as_ref()
+      .map(|args| {
+        args
+          .iter()
+          .map(|(k, v)| format!(r#"{}: {}"#, k, v.render_graphql(ctx).escape_default()))
+          .collect::<Vec<_>>()
+          .join(", ")
+      })
+      .map(|args| format!("{}({})", self.operation_name, args))
+      .unwrap_or(self.operation_name.clone());
+
+    let graphql_query = format!(r#"{{ "query": "{operation_type} {{ {operation} {selection_set} }}" }}"#);
```

Comment:
Instead of hand-crafting a GraphQL query, we should use:

`async_graphql_parser::types::executable::ExecutableDocument`

Reply:
There are some rough edges in using this:
- the main purpose for the struct is to parse query to the server it's little harder to convert it to string to pass further. yeah, there is a method to [stringify](https://docs.rs/async-graphql/latest/async_graphql/extensions/struct.ExtensionContext.html#method.stringify_execute_doc) it, by it's main purpose for logging and it could mess with the output to hide sensitive data for example
- in order to create ExecutableDocument we need to do different types of conversion into types that used in this type to at the end convert it back to string. It could be simpler and faster just convert it to string. And I think we can't do this upfront as selection_set of fields could be different. 
- also in current implementation we couldn't easily use variables here because to use it we need to know different arguments types to specify query like this `query(arg: String, arg2: Int) { user(id: $arg2) { name } }` and it's not a very straightforward. Previously, it was done by using upstream instrospection, but I removed it and started to pass arguments explicitly with `path_graphql`

Overall, I'm not sure it's worth the effort